beacon_node/lighthouse_network/src/peer_manager/mod.rs

```diff
@@ -2051,4 +2051,149 @@ mod tests {
+
+        impl Arbitrary for AttestationSubnetId {
+            fn arbitrary(g: &mut Gen) -> Self {
+                let len = <E as EthSpec>::SubnetBitfieldLength::to_usize();
+                let id_range = (0..len).collect::<Vec<usize>>();
+                AttestationSubnetId(*g.choose(&id_range).unwrap())
+            }
+        }
+
+        #[derive(Clone, Debug)]
+        struct SyncCommitteeSubnetId(usize);
+
+        impl Arbitrary for SyncCommitteeSubnetId {
+            fn arbitrary(g: &mut Gen) -> Self {
+                let len = <E as EthSpec>::SyncCommitteeSubnetCount::to_usize();
+                let id_range = (0..len).collect::<Vec<usize>>();
+                SyncCommitteeSubnetId(*g.choose(&id_range).unwrap())
+            }
+        }
+
+        #[quickcheck]
+        fn prune_excess_peers(peer_conditions: Vec<PeerCondition>) -> TestResult {
+            let target_peer_count = DEFAULT_TARGET_PEERS;
+            let rt = Runtime::new().unwrap();
+
+            rt.block_on(async move {
+                let mut peer_manager = build_peer_manager(target_peer_count).await;
+
+                // Create peers based on the randomly generated conditions.
+                for condition in &peer_conditions {
+                    let peer = PeerId::random();
+                    let mut attnets = crate::types::EnrAttestationBitfield::<E>::new();
+                    let mut syncnets = crate::types::EnrSyncCommitteeBitfield::<E>::new();
+
+                    match condition.direction {
+                        Direction::Incoming => {
+                            peer_manager.inject_connect_ingoing(
+                                &peer,
+                                "/ip4/0.0.0.0".parse().unwrap(),
+                                None,
+                            );
+                        }
+                        Direction::Outgoing => {
+                            peer_manager.inject_connect_outgoing(
+                                &peer,
+                                "/ip4/0.0.0.0".parse().unwrap(),
+                                None,
+                            );
+                        }
+                    }
+
+                    for id in &condition.attestation_subnets {
+                        attnets.set(id.0, true).unwrap();
+                    }
+
+                    for id in &condition.sync_committee_subnets {
+                        syncnets.set(id.0, true).unwrap();
+                    }
+
+                    let metadata = crate::rpc::MetaDataV2 {
+                        seq_number: 0,
+                        attnets,
+                        syncnets,
+                    };
+
+                    let mut peer_db = peer_manager.network_globals.peers.write();
+                    let peer_info = peer_db.peer_info_mut(&peer).unwrap();
+                    peer_info.set_meta_data(MetaData::V2(metadata));
+                    peer_info.set_gossipsub_score(condition.gossipsub_score);
```

Comment:
Unlike [`set_lighthouse_score`](https://github.com/sigp/lighthouse/blob/df40700ddd2dcc3c73859cc3f8e315eab899d87c/beacon_node/lighthouse_network/src/peer_manager/peerdb/score.rs#L182), [`set_gossipsub_score`](https://github.com/sigp/lighthouse/blob/df40700ddd2dcc3c73859cc3f8e315eab899d87c/beacon_node/lighthouse_network/src/peer_manager/peerdb/score.rs#L224) doesn't induce `update_state()`. So `set_gossipsub_score` should be called before `add_to_score` here to update the score accurately.

Reply:
> [set_gossipsub_score](https://github.com/sigp/lighthouse/blob/df40700ddd2dcc3c73859cc3f8e315eab899d87c/beacon_node/lighthouse_network/src/peer_manager/peerdb/score.rs#L224) doesn't induce update_state()

Is that intentional?

Reply:
I dont think its intentional. We should add `update_state()` to `set_gossipsub_score()`